### PR TITLE
Perf. improvement for `CuStateVecCircuitSimulator::observe`

### PR DESCRIPTION
**Pull request by 1tnguyen**
### Description

Currently, we don't use the batched Pauli expectation API for multi-term `spin_op` expectation calculation. 

This API is substantially faster than applying change-of-basis gates then observing `<ZZZ..ZZ>` for each term.

Hence, this PR modifies `custatevec` backend to use batched Pauli expectation API by default for deterministic `cudaq::observe` (no shots)

(1) Change `CircuitSimulator::observe` to return `observe_result`, which can encapsulate the final expectation value as well as individual terms' expectation values.

(2) Use `custatevec` batched API and populate the `observe_result` accordingly.

(3) Add a backward compatibility test for the term-by-term mode.

### CODE REVIEW

File: runtime/nvqir/custatevec/CuStateVecCircuitSimulator.cu [hunk 1]
```diff
@@ -437,34 +437,80 @@ public:
...
+    }
 
-    flushGateQueue();
+    // Use batched custatevecComputeExpectationsOnPauliBasis to compute all term
+    // expectation values in one go
+    uint32_t nPauliOperatorArrays = op.num_terms();
+    // Stable holders of vectors since we need to send vectors of pointers to
+    // custatevec
+    std::deque<std::vector<custatevecPauli_t>> pauliOperatorsArrayHolder;
+    std::deque<std::vector<int32_t>> basisBitsArrayHolder;
```
Comment (schweitzpgi) [inline runtime/nvqir/custatevec/CuStateVecCircuitSimulator.cu:457]: Why use a `deque` for these?

**Review by schweitzpgi — COMMENTED** [review]
LGTM.

**Review by bmhowe23 — COMMENTED** [review]
I suspect we need some docs for the `CUDAQ_OBSERVE_FROM_SAMPLING` environment variable.

**Comment by amccaskey** [comment]
@1tnguyen seems like we should also update `CircuitSimulator::shouldObserveFromSampling()` to return `false` by default. What do you think? Should this new approach be the default? Is it always faster?

**Comment by 1tnguyen** [comment]
> @1tnguyen seems like we should also update `CircuitSimulator::shouldObserveFromSampling()` to return `false` by default. What do you think? Should this new approach be the default? Is it always faster?

Yes, it's always faster: a lot faster for large Hamiltonians since the time to run change-of-basis gates and then uncompute them is not insignificant. 

There is an issue with making it the default. Currently, we expect that `cudaq::observe` does return term-by-term results like in this [test](https://github.com/NVIDIA/cuda-quantum/blob/c38899424b679132b6019621b5320b855541ee8d/unittests/integration/observe_result_tester.cpp#L84).
Hence, we need to make an additional change to make `CircuitSimulator::observe` able to return term-by-term results if it can (like in this case, we perform batch exp-val computation and do have all the individual results). Then, we can make `CuStateVecCircuitSimulator::shouldObserveFromSampling` return true by default.

What do you think?

**Comment by amccaskey** [comment]
> > @1tnguyen seems like we should also update `CircuitSimulator::shouldObserveFromSampling()` to return `false` by default. What do you think? Should this new approach be the default? Is it always faster?
> 
> Yes, it's always faster: a lot faster for large Hamiltonians since the time to run change-of-basis gates and then uncompute them is not insignificant.
> 
> There is an issue with making it the default. Currently, we expect that `cudaq::observe` does return term-by-term results like in this [test](https://github.com/NVIDIA/cuda-quantum/blob/c38899424b679132b6019621b5320b855541ee8d/unittests/integration/observe_result_tester.cpp#L84). Hence, we need to make an additional change to make `CircuitSimulator::observe` able to return term-by-term results if it can (like in this case, we perform batch exp-val computation and do have all the individual results). Then, we can make `CuStateVecCircuitSimulator::shouldObserveFromSampling` return true by default.
> 
> What do you think?

I say go for it - make that change here in this PR.

**Comment by 1tnguyen** [comment]
This PR is ready for (re-)review. Ping reviewers... :-)

File: unittests/CMakeLists.txt [hunk 2]
```diff
@@ -120,7 +120,30 @@ if (CUSTATEVEC_ROOT AND CUDA_FOUND)
     cudaq-platform-mqpu
     nvqir-custatevec-fp64
     gtest_main)
-    gtest_discover_tests(test_mqpu PROPERTIES LABELS "gpu_required")
+  gtest_discover_tests(test_mqpu PROPERTIES LABELS "gpu_required")
+
+  # Test CUDAQ_OBSERVE_FROM_SAMPLING=ON mode.
```
Comment (bmhowe23) [inline unittests/CMakeLists.txt:125]: Just to confirm ... we also have tests w/ CUDAQ_OBSERVE_FROM_SAMPLING disabled, right? I assume all the prior custatevec tests will effectively be running with CUDAQ_OBSERVE_FROM_SAMPLING effectively off.

File: unittests/backends/qpp_observe/QppObserveTester.cpp [hunk 3]
```diff
@@ -27,7 +27,7 @@ CUDAQ_TEST(QPPBackendTester, checkBackendObserve) {
                      .21829 * z(0) - 6.125 * z(1);
 
   auto expVal = qpp.observe(h);
-  EXPECT_NEAR(expVal.expectationValue.value(), -1.74, 1e-2);
+  EXPECT_NEAR(expVal.expectation(), -1.74, 1e-2);
```
Comment (bmhowe23) [inline unittests/backends/qpp_observe/QppObserveTester.cpp:30]: Is this an interface change that is visible to the user?

**Review by bmhowe23 — COMMENTED** [review]
This will be a very nice optimization. Are all the interface changes invisible to the end user?

File: unittests/CMakeLists.txt (hunk 2 above)
Comment (1tnguyen) [reply unittests/CMakeLists.txt:125]: yes, that's right. All the prior tests (using `observe` without shots) should be running in the `CUDAQ_OBSERVE_FROM_SAMPLING=OFF`  (new code path using batched observe).

File: unittests/backends/qpp_observe/QppObserveTester.cpp (hunk 3 above)
Comment (1tnguyen) [reply unittests/backends/qpp_observe/QppObserveTester.cpp:30]: This is not visible to regular users unless they retrieve the `CircuitSimulator` instance in user code, e.g., via `cudaq::getUniquePluginInstance` and interact with the simulator directly.

**Review by bmhowe23 — APPROVED** [review]
👍

**Pull request merged**